src/data/polynomial/lifts.lean

```diff
@@ -0,0 +1,332 @@
+lemma lifts_sum_mem {ι : Type w} {t : finset ι} {g : ι → (polynomial S)} :
+  (∀ (x : ι), x ∈ t → lifts f (g x)) → lifts f (∑ (x : ι) in t, g x) :=
+by simp_rw [lifts_iff]; exact subsemiring.sum_mem (ring_hom.of (map f)).srange
+
+/--If `p` lifts then is leading monomial lifts. -/
+lemma lead_monom_lifts_of_lifts {p : polynomial S} (h : lifts f p) :
+  lifts f ((C p.leading_coeff) * polynomial.X ^ p.nat_degree) :=
+begin
+  obtain ⟨q, hq⟩ := h,
+  rw leading_coeff,
+  use (C (q.coeff p.nat_degree)) * polynomial.X ^ p.nat_degree,
+  nth_rewrite 2 ←hq,
+  simp only [map_C, coeff_map, map_pow, map_X, map_mul]
+end
+
+/--If `p` lifts then `p.erase_lead` lifts. -/
+lemma erase_lead_lifts_of_lifts {p : polynomial S} (h : lifts f p) : lifts f p.erase_lead :=
+begin
+  cases erase_lead_nat_degree_lt_or_erase_lead_eq_zero p with hdeg hzero,
+  { obtain ⟨q, hq⟩ := h,
+    have hcoeff : ∀ n ∈ p.erase_lead.support, f (q.coeff n) = p.erase_lead.coeff n,
+    { intros n hn,
+      rw [erase_lead_coeff_of_ne n (ne_nat_degree_of_mem_erase_lead_support hn), ← hq, coeff_map] },
+    use (∑ (i : ℕ) in p.erase_lead.support, (C (q.coeff i)) * polynomial.X ^ i),
+    simp [map_sum, map_C, map_pow, map_X, map_mul],
+    conv_lhs { apply_congr,
+               skip,
+               simp only [hcoeff, H] },
+    nth_rewrite 1 [as_sum_support p.erase_lead] },
+  { use 0; simp only [hzero, map_zero] },
+end
+
+section lift_deg
+
+/--`lifts_deg f p` means that `p` can be lifted to a polynomial of the same degree. -/
+def lifts_deg (f : R →+* S) (p : polynomial S) : Prop :=
+  ∃ (q : polynomial R), map f q = p ∧ q.degree = p.degree
+
+lemma lifts_deg_of_supp_le_one {p : polynomial S} (hs : p.support.card < 2) (hl : lifts f p) :
+  lifts_deg f p :=
+begin
+  have h : p.support.card = 0 ∨ p.support.card = 1 := by omega,
+  cases h with h0 h1,
+  { simp only [finsupp.support_eq_empty, finset.card_eq_zero] at h0,
+    use 0,
+    simp only [h0, degree_zero, eq_self_iff_true, and_self, map_zero] },
+  have hzero : p ≠ 0 := by rw [←nonempty_support_iff, ←finset.card_pos, h1]; exact zero_lt_one,
+  replace hs := C_mul_X_pow_eq_self (le_of_eq h1),
+  rw [← hs, lifts_deg],
+  obtain ⟨q, hq⟩ := hl,
+  use (C(q.coeff p.nat_degree) * X ^ p.nat_degree),
+  have hcoeff : f (q.coeff p.nat_degree ) = p.coeff p.nat_degree,
+  { rw [← hq, coeff_map] },
+  split,
+  { simp only [leading_coeff, hcoeff, map_C, map_pow, map_X, map_mul] },
+  have pcoeff : p.leading_coeff ≠ 0 := by intro ha; exact hzero (leading_coeff_eq_zero.1 ha),
+  have qcoeff : q.coeff p.nat_degree ≠ 0,
+  { intro habs,
+    rw [habs, ←leading_coeff, ring_hom.map_zero] at hcoeff,
+    exact pcoeff hcoeff.symm },
+  simp only [qcoeff, pcoeff, ne.def, not_false_iff, degree_monomial]
+end
+
+/--A polynomial lifts if and only if it can be lifted to a polynomial of the same degree. -/
+lemma lifts_iff_lifts_deg (f : R →+* S) (p : polynomial S) : lifts_deg f p ↔ lifts f p :=
```

Comment:
The sides of the name and the sides of the statement are swapped.

Reply:
Corrected, thanks.